# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [31]:
# Loading required packages

import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set_style('darkgrid')

# Obtaining the current working directory
import os
os.chdir('C:\\Users\\omkar\\OneDrive\\Springboard\\Assignments for submission\\Guided Capstone\\data')

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [32]:
df = pd.read_csv('step3_output.csv')
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     0       0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              505.0       60.000000        163.000000       82.0   
4             1500.0      174.873239        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3             0.0         1  
4             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [33]:
df = pd.concat([df.drop('state',axis=1),pd.get_dummies(df['state'])],axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [35]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [36]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [37]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [38]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [39]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [40]:
from sklearn.metrics import explained_variance_score,mean_absolute_error
from math import sqrt
exp_var_score = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)

print(exp_var_score)
print(mae)

0.6627317182217789
5.987484761578867


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [41]:
print(lm.intercept_)

13.855544612383959


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [47]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff_1 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff_1.sort_values(by = 'Coefficient', ascending = False)

Coefficient
Tennessee          2.309271e+01
North Carolina     1.572401e+01
Wyoming            1.137267e+01
South Dakota       1.125029e+01
Massachusetts      1.120385e+01
New Mexico         1.102649e+01
Idaho              1.023223e+01
Connecticut        9.527853e+00
Arizona            9.195448e+00
Montana            8.559506e+00
New Hampshire      8.182443e+00
Utah               7.413031e+00
Virginia           7.268766e+00
West Virginia      6.972328e+00
Oregon             6.369022e+00
Vermont            5.634878e+00
Maine              5.020133e+00
Iowa               4.592676e+00
Pennsylvania       4.583998e+00
Colorado           4.499927e+00
Ohio               3.914001e+00
Washington         3.580623e+00
California         3.464988e+00
Indiana            3.290017e+00
Alaska             2.892340e+00
Nevada             2.801459e+00
Illinois           2.060312e+00
Wisconsin          1.755878e+00
Michigan           1.390984e+00
clusters           1.260430e+00
Minnesota          1.165138e+00
New York           1.044015e+00
LongestRun_mi      1.034010e+00
triple             9.143330e-01
surface            8.748473e-01
AdultWeekday       6.508355e-01
total_chairs       4.903591e-01
TerrainParks       3.783748e-01
double             3.409370e-01
fastQuads          3.145855e-01
quad               2.046491e-01
Missouri           1.725714e-01
Runs               1.219764e-01
yearsOpen          3.492014e-02
NightSkiing_ac     2.848913e-02
daysOpenLastYear   2.239597e-02
projectedDaysOpen  1.890674e-02
averageSnowfall    1.819468e-03
SkiableTerrain_ac  1.573426e-03
vertical_drop      1.114143e-03
Snow Making_ac     8.214082e-04
summit_elev        6.853086e-04
fastEight          5.547833e-12
trams              4.920897e-12
fastSixes          6.518119e-13
Maryland           5.329071e-15
New Jersey         1.776357e-15
Rhode Island       0.000000e+00

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [51]:
# Importing the pre-processed data again
df_pre = pd.read_csv('step3_output.csv')
# Drop the state variable
df_pre.drop('state',axis=1,inplace=True)

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_pre.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_pre['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 


# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [52]:
#all second model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model_2 = lm.fit(X_train,y_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model_2.predict(X_test)

from sklearn.metrics import explained_variance_score,mean_absolute_error
from math import sqrt
exp_var_score = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)

print(exp_var_score)
print(mae)

0.5559565165902218
6.8529906862407275


In [55]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff_2 = pd.DataFrame(abs(model_2.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending = False)
coeff_2

Coefficient
clusters           3.924873e+00
surface            1.082570e+00
triple             9.208312e-01
AdultWeekday       6.960583e-01
total_chairs       3.875508e-01
quad               3.772698e-01
LongestRun_mi      3.378559e-01
double             2.787507e-01
fastQuads          1.067314e-01
daysOpenLastYear   9.431953e-02
Runs               9.146455e-02
projectedDaysOpen  3.011355e-02
TerrainParks       2.683867e-02
NightSkiing_ac     1.842024e-02
averageSnowfall    1.838393e-02
yearsOpen          9.224870e-03
Snow Making_ac     3.265422e-03
vertical_drop      2.436503e-03
summit_elev        1.056583e-03
SkiableTerrain_ac  3.775805e-04
fastSixes          1.942890e-15
fastEight          1.887379e-15
trams              3.469447e-18

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [57]:
# Importing the pre-processed data again
df_3 = pd.read_csv('step3_output.csv')
# Drop the state variable
df_3.drop(['state','summit_elev'],axis=1,inplace=True)

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_3.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_3['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 


# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [58]:
#all second model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model_3 = lm.fit(X_train,y_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model_3.predict(X_test)

from sklearn.metrics import explained_variance_score,mean_absolute_error
from math import sqrt
exp_var_score = explained_variance_score(y_test,y_pred)
mae = mean_absolute_error(y_test,y_pred)

print(exp_var_score)
print(mae)

0.5413080297878097
7.275361831392759


In [59]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff_3 = pd.DataFrame(abs(model_3.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending = False)
coeff_3

Coefficient
clusters           2.989593e+00
triple             8.544500e-01
surface            8.339838e-01
AdultWeekday       6.790482e-01
total_chairs       4.476521e-01
double             4.311239e-01
TerrainParks       2.448850e-01
fastQuads          2.007948e-01
quad               1.968568e-01
LongestRun_mi      7.669402e-02
daysOpenLastYear   6.671560e-02
Runs               6.433675e-02
averageSnowfall    2.740369e-02
NightSkiing_ac     2.354828e-02
projectedDaysOpen  6.732339e-03
SkiableTerrain_ac  5.476078e-03
Snow Making_ac     3.858733e-03
vertical_drop      3.275773e-03
yearsOpen          1.927092e-03
trams              9.763024e-15
fastSixes          2.109424e-15
fastEight          1.221245e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.66 |5.98  |-|
| Model 2. | 0.55|6.85 |'state'|
| Model 3. | 0.54 |7.27 |'state','summit_elev','base_elev'|

Model Selection: